Backend/services/user/index.js

```diff
@@ -1,44 +1,150 @@
-const express = require('express')
-const router = express.Router()
+require("dotenv").config();
+
+const express = require("express");
+const router = express.Router();
+
+const Pool = require("pg").Pool;
+
+const pool = new Pool({
+  user: process.env.DB_USER,
+  host: process.env.DB_HOST,
+  database: process.env.DATABASE,
+  password: process.env.DB_PASSWORD,
+  port: process.env.PORT,
+});
+
+const jwt = require("jsonwebtoken");
+const bcrypt = require("bcrypt");
+
+const auth = require("../auth");
 
 /**
  * Returns the user list.
- * Observe good security practices such as removing the password field 
+ * Observe good security practices such as removing the password field
  * from the response object.
- * 
+ *
  * If :id is given, the user with the given id will be returned.
  */
-router.get('/', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+
+router.get("/", auth.authenticateToken, (request, response) => {
+  pool.query(
+    "SELECT ID, users.firstname, users.lastname, users.username FROM users",
+    (error, result) => {
+      if (error) {
+        response.status(500).send({
+          error: error,
+        });
+      }
+      response.status(200).json(result.rows);
+    }
+  );
+});
 
 /**
  * Returns the user with the given id.
  */
-router.get('/:id', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.get("/:id", auth.authenticateToken, (request, response) => {
+  console.log(auth.authenticateToken);
+  const id = parseInt(request.params.id);
+  pool.query(
+    "SELECT ID, users.firstname, users.lastname, users.username FROM users WHERE id = $1",
+    [id],
+    (error, result) => {
+      if (error) {
+        response.status(500).send({
+          error: error,
+        });
+      }
+      if (result.rows[0].username === request.user.username) {
+        return response.status(200).json(result.rows[0]);
+      }
+      return response.status(200).send("not allowed");
+    }
+  );
+});
 
 /**
  * Registers a new user.
  */
-router.post('/register', (request, response) => {
+router.post("/register", async (request, response) => {
+  const { firstName, lastName, username, password } = request.body;
+
+  try {
+    const salt = await bcrypt.genSalt();
+    const hashedPassword = await bcrypt.hash(password, salt);
+    if (username) {
+      pool.query(
+        `SELECT users.username FROM users WHERE users.username = $1`,
+        [username],
+        function (error, result) {
+          if (error) {
+            response.status(500).send({
+              error: error,
+            });
+          }
+          if (result.rows.length < 1) {
```

Comment:
Make use of the `UNIQUE` schema constraint for `username` so that you don't have to do this.

For example, `username text UNIQUE NOT NULL`